homeassistant/components/motion_blinds/__init__.py

```diff
@@ -31,8 +38,26 @@ async def async_setup_entry(
     host = entry.data[CONF_HOST]
     key = entry.data[CONF_API_KEY]
 
+    # Create multicast Listener
+    multicast = hass.data[DOMAIN].setdefault(
+        KEY_MULTICAST_LISTENER,
+        MotionMulticast(),
+    )
+
+    if len(hass.data[DOMAIN]) == 1:
+        # start listening for local pushes (only once)
+        await hass.async_add_executor_job(multicast.Start_listen)
+
+        # register stop callback to shutdown listening for local pushes
+        def stop_motion_multicast(event):
+            """Stop multicast thread."""
+            _LOGGER.debug("Shutting down Motion Listener")
+            multicast.Stop_listen()
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_motion_multicast)
```

Comment:
My read of Stop_listen is that it is not idempotent and if called twice it might throw as thread_ is set to None the first time through.    I am trying to reason about that there is both a stop call here, and one when all config entries are unloaded.

Are all config entries unloaded on home assistant stop? I was curious if maybe that check alone is sufficient to handle all cleanup, or if something more intricate is needed for clean shutdown.

Reply:
No config entries are not unloaded on home assistant stop, that means that the multicast.Stop_listen() is only called once.

However you are right that if you remove the motion_blinds integration, the multicast.Stop_listen() is called and if you then stop HomeAssistant it is called again, indeed generating an error. However that error will only occur once since the motion_blinds is then not setup again at a start of HomeAssistant.

I just made a commit to upstream motion_blinds to prevent errors if multicast.Stop_listen() is called multiple times.
That fix will end up in the next release of motion_blinds (probebly before all of the HomeAssistant PRs get merged).

Reply:
Making it idempotent sounds good to me.

Reply:
Making shutdown idempotent sounds good.